### PR TITLE
:recycle: Address low priority logging issue

### PR DESCRIPTION
**Pull request by mrsegen**
This seeks to address an issue with error logging. There are also _some_ increases tests and some proposed tests as a conversation piece. 

I was seeing messages like the following when there is a sell signal for a token that isn't yet held in a BinanceUS account:
```
Unable to fetch LTC ticker balance (may be 0) -> TypeError: Cannot read property 'free' of undefined.
Failed to close LTC/USD position.
Failed to open position on LTC/USD.
```

As one of the first users of this with BinanceUS - in case this would be a question, I can confirm that I _was_ able to do manually triggered buy and sell signals:
```
[{"message":"exchanges|BinanceUS/[STUB] - WAVES ticker balance successfully fetched. -> {\"coin\":\"WAVES\",\"free\":\"0.36000000\",\"total\":0.36}","level":"debug","timestamp":"2021-09-12T03:13:47.814Z"},
{"message":"trading|Calculated 0.35961462389214177 WAVES/USD equivalent for 11.00 $US.","level":"debug","timestamp":"2021-09-12T03:13:47.815Z"},
{"message":"trading|Calculated 0.35961462389214177 WAVES/USD equivalent for 11.00 $US.","level":"debug","timestamp":"2021-09-12T03:13:47.815Z"},
{"message":"trading|BinanceUS/[STUB] - $$$ Closing 0.35961462389214177 WAVES/USD of open position / available balance (~ 11.00 $US).","level":"info","timestamp":"2021-09-12T03:13:48.754Z"},
{"message":"trading|Trade [omitted-uuid] executed in 2564 ms.","level":"info","timestamp":"2021-09-12T03:13:48.755Z"}]
```

===============

Notably, there are a number of `todo`ed tests:
* Some of these may be hard to write without some dependency injection, which seemed a bit much to try without some thought/discussion. 
* I suppose some of them would offer little value, but I just wrote them as a way to help myself better understand what was at play in various functions. Perhaps they could just be replaced with function documentation.

### CODE REVIEW

**Comment by mrsegen** [comment]
I have got to acknowledge that, so far, this doesn't actually address handling of the error that arose. Rather, it breaks up code so that a "failure to open an order" error doesn't get logged when it was more of a "failure to close a position."

Actually identifying the precise location and cause of the error (cannot find a property of an undefined variable) may require some debugging. I have done some initial exploration of how to enable sandbox mode for an exchange on a separate branch but haven't yet made substantial progress on it.

File: src/services/exchanges/base/base.exchange.service.ts
Comment (mrsegen) [inline src/services/exchanges/base/base.exchange.service.ts:286]: My thinking here was to distinguish between "open position" and "open order".

File: src/services/exchanges/base/base.exchange.service.ts
Comment (thibaultyou) [inline src/services/exchanges/base/base.exchange.service.ts:220]:
```suggestion
      if (!balance) {
```

File: src/services/exchanges/base/base.exchange.service.ts
Comment (thibaultyou) [reply src/services/exchanges/base/base.exchange.service.ts:286]: yes why not :)

File: src/services/exchanges/base/composite.exchange.service.ts
Comment (thibaultyou) [inline src/services/exchanges/base/composite.exchange.service.ts:25]:
```suggestion
      if (!balance) {
```

File: src/services/exchanges/base/composite.exchange.service.ts
Comment (thibaultyou) [inline src/services/exchanges/base/composite.exchange.service.ts:26]: I'm not a big fan of putting strings outside constants, we should find a workaround for this.

File: src/services/exchanges/base/composite.exchange.service.ts
Comment (thibaultyou) [inline src/services/exchanges/base/composite.exchange.service.ts:36]: Same as above here, even if I get the point of using it.

File: src/services/exchanges/base/spot.exchange.service.ts
Comment (thibaultyou) [inline src/services/exchanges/base/spot.exchange.service.ts:36]:
```suggestion
      if (!balance) {
```

File: src/services/exchanges/base/spot.exchange.service.ts
Comment (thibaultyou) [inline src/services/exchanges/base/spot.exchange.service.ts:47]: Same as above.

**Review by thibaultyou — CHANGES_REQUESTED** [review]
We just need to find a workaround to make a distinction between errors without passing strings to constructors. Once this is done I'll test if there's no regressions on my side and merge it to master.

Thanks for putting time and effort into this. 🙏

**Review by thibaultyou — APPROVED** [review]

**Comment by thibaultyou** [comment]
Thanks for the update.

Let me test this once I've some free time and I'll merge this into master if you're done with this PR.

🙏

**Comment by mrsegen** [comment]
@thibaultyou yep, I'm done with the changes I was planning

**Comment by thibaultyou** [comment]
Just tested it, it's good for me.
I'll merge it now to master and update the Docker image once I'm logged on a non-ARM architecture.

Thank you for your contribution 🙏 .

**Comment by thibaultyou** [comment]
The Docker image has been updated.

**Pull request merged**
